### PR TITLE
gh-116738: Make `_codecs` module thread-safe

### PR DESCRIPTION
**Pull request by swtaarrs**
The module itself is a thin wrapper around calls to functions in `Python/codecs.c`, so that's where the meaningful changes happened:
- Move codecs-related state that lives on `PyInterpreterState` to a struct declared in `pycore_codecs.h`.
- In free-threaded builds, add two mutexes to `codecs_state`, one to synchronize operations on `search_path`, and one to synchronize initialization.
- `_PyCodecRegistry_Init()` is now `_PyCodecRegistry_EnsureInit()`, and is called unconditionally to ensure initialization, rather than only when initialization is needed. When it returns 0 (as opposed to -1 to indicate an error), the `PyObject *` members of `codecs_state` can be read without further synchronization.
- Operations that mutate `codecs.search_path` must be performed while holding `codecs.search_path_mutex`. This allows `PyCodec_Unregister()` to search for and delete a specific item from the list. Because `search_path_mutex` is used as a normal mutex and not a critical section, we must be extremely careful with operations called while holding it.

- Issue: gh-116738

### CODE REVIEW

File: Python/codecs.c [hunk 1]
```diff
@@ -1463,45 +1488,74 @@ static int _PyCodecRegistry_Init(void)
...
-
-    interp->codec_search_path = PyList_New(0);
-    if (interp->codec_search_path == NULL) {
-        return -1;
     }
 
-    interp->codec_search_cache = PyDict_New();
-    if (interp->codec_search_cache == NULL) {
-        return -1;
+    PyObject *search_path = NULL, *search_cache = NULL, *error_registry = NULL;
```
Comment (colesbury) [inline Python/codecs.c]: A few suggestions:

1) Let's move the `init_mutex` acquisition up. It should be safe to allocate `search_path`, etc. within the mutex
2) Let's use seq-cst atomics. Although I think that acq/rel is probably sufficient here, let's use seq-cst outside of places where there's a good perf motivation for using weaker atomics. I think it's also fine (and easier to read & maintain) if we just use the `_Py_atomic` functions here unconditionally instead of adding additional "ft" wrappers.
3) We should only skip the import of `encodings` if we've already successfully imported that module. It's OK to have multiple threads try to import `encodings` concurrently. The import subsystem has internal locking to handle that.

File: Python/codecs.c [hunk 2]
```diff
@@ -55,22 +66,33 @@ int
 PyCodec_Unregister(PyObject *search_function)
 {
     PyInterpreterState *interp = _PyInterpreterState_GET();
-    PyObject *codec_search_path = interp->codec_search_path;
-    /* Do nothing if codec_search_path is not created yet or was cleared. */
-    if (codec_search_path == NULL) {
+    /* Do nothing if codec data structures are not created yet. */
+    if (FT_ATOMIC_LOAD_INT_ACQUIRE(interp->codecs.initialized) == 0) {
```
Comment (colesbury) [inline Python/codecs.c]: I think we should do this check under the `init_mutex` lock.

**Review by colesbury — COMMENTED** [review]
This looks pretty good. Two comments below

File: Python/codecs.c (hunk 1 above)
Comment (mpage) [reply Python/codecs.c]: I think it might be marginally simpler, but why not use a `_PyOnceFlag` for initialization?

Comment (swtaarrs) [reply Python/codecs.c]: Oh yeah, I remember now that we had talked about `_PyOnceFlag`. I'll give it a try.

Comment (colesbury) [reply Python/codecs.c]: I was going to suggest `_PyOnceFlag`, but I think we'll want to synchronize `PyCodec_Unregister` with initialization and that's harder to do with a `_PyOnceFlag`.

If you get it to work well with `_PyOnceFlag`, then go for it.

Comment (swtaarrs) [reply Python/codecs.c]: A straightforward conversion to `_PyOnceFlag` would mean that `PyCodec_Unregister()` triggers initialization if it hasn't already happened (as opposed to now, where it just returns early). If initialization didn't include something as heavy as importing `encodings`, that might be a reasonable tradeoff, but I think it makes sense to stick with the manual synchronization in this case.

Comment (swtaarrs) [reply Python/codecs.c]: > We should only skip the import of encodings if we've already successfully imported that module.

How strictly should I follow this? The old version of this code would only try to import `encodings` once. If that failed, that specific call to `_PyCodecRegistry_Init()` would return -1, but subsequent calls into this file would think initialization was complete. [This line](https://github.com/python/cpython/blob/6258844c27e3b5a43816e7c559089a5fe0a47123/Python/codecs.c#L1505) after the import is misleading; nothing reads `codecs_initialized` in the old code.

I matched that behavior (for `encodings`, anyway. the new code will avoid setting any of the interpreter-global state if we fail to create part of it). If you're suggesting that I instead have subsequent calls to `_PyCodecRegistry_EnsureInit()` retry the import if it failed last time, I can do that, but it will require another piece of state tracking whether the import succeeded. It would be a visible behavioral change, although only in error cases.

Comment (colesbury) [reply Python/codecs.c]: OK, yeah maybe don't change it. I'm looking at how this is actually called and it happens pretty early in interpreter initialization:

```
init_interp_main -> _PyUnicode_InitEncodings -> init_fs_encoding -> config_get_codec_name -> _PyCodec_Lookup -> _PyCodecRegistry_Init -> ... -> PyCodec_Register
```

I think we could simplify this all by explicitly initializing the codec registry instead of sort-of-lazily initializing it, but that would be a more involved change than you PR.

File: Python/codecs.c (hunk 2 above)
Comment (swtaarrs) [reply Python/codecs.c]: What would that protect against? I should probably change it to `if (initialized != 1) return` rather than `if (initialized == 0) return`, but if this check needs to be done while holding the lock, [this one](https://github.com/python/cpython/pull/117530/files#diff-a22bcb6c26007e5597e92a6abb0a27ce987072fcfa7220623c869044c3097784R1491-R1493) should be too.

Comment (colesbury) [reply Python/codecs.c]: The suggestion was more motivated about being easier to reason about than a correctness issue.

But I don't think it matters much because we don't actually lazily initialize codecs.

File: Python/codecs.c [hunk 3]
```diff
@@ -11,6 +11,7 @@ Copyright (c) Corporation for National Research Initiatives.
 #include "Python.h"
 #include "pycore_call.h"          // _PyObject_CallNoArgs()
 #include "pycore_interp.h"        // PyInterpreterState.codec_search_path
+#include "pycore_lock.h"          // PyMutex
 #include "pycore_pyerrors.h"      // _PyErr_FormatNote()
 #include "pycore_pystate.h"       // _PyInterpreterState_GET()
 #include "pycore_ucnhash.h"       // _PyUnicode_Name_CAPI
```
Comment (colesbury) [inline Python/codecs.c:17]: Please update the out of date comment below

File: Python/pystate.c [hunk 4]
```diff
@@ -838,9 +838,10 @@ interpreter_clear(PyInterpreterState *interp, PyThreadState *tstate)
     }
 
     PyConfig_Clear(&interp->config);
-    Py_CLEAR(interp->codec_search_path);
-    Py_CLEAR(interp->codec_search_cache);
-    Py_CLEAR(interp->codec_error_registry);
+    Py_CLEAR(interp->codecs.search_path);
+    Py_CLEAR(interp->codecs.search_cache);
+    Py_CLEAR(interp->codecs.error_registry);
+    interp->codecs.initialized = 0;
```
Comment (colesbury) [inline Python/pystate.c]: I'd refactor this into something like `_PyCodec_Fini()` and move it to `codecs.c`

File: Objects/object.c [hunk 5]
```diff
@@ -473,7 +473,6 @@ void
 PyObject_CallFinalizer(PyObject *self)
 {
     PyTypeObject *tp = Py_TYPE(self);
-
```
Comment (colesbury) [inline Objects/object.c]: Remove unrelated change

**Review by colesbury — COMMENTED** [review]
This looks good. A few comments below.

File: Python/codecs.c (hunk 3 above)
Comment (swtaarrs) [reply Python/codecs.c:17]: I deleted it except for the part about registering the first codec search function, which I moved to the comment I added in `pycore_codecs.h` (the part about ignoring `ImportError`s was already false before my changes).

**Comment by swtaarrs** [comment]
This should be ready to go as of my last commit - let me know if any other changes are needed.

**Review by colesbury — APPROVED** [review]
